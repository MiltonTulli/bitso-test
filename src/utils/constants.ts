export const MATRIX_SIZES = ['5x5', '10x5', '10x10', '15x10', '15x15', '20x20'];
export const IS_MOBILE_BREAK = 600;
