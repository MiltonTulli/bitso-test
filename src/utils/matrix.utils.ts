import { ICell, IMatrix } from '../interfaces';

// export utils -----------------
/**
 * @function createMatrix
 * @description Create a matrix (array of arrays) given cols & rows
 * @param {Number} cols - Num of columns (x)
 * @param {Number} rows - Num of rows (y)
 */
export const createMatrix = (cols: number, rows: number): ICell[][] => {
  let matrix: ICell[][] = [];
  for (let y = 0; y < rows; y++) {
    matrix[y] = [];
    for (let x = 0; x < cols; x++) {
      matrix[y][x] = {
        id: `${x}-${y}`,
        active: false
      };
    }
  }
  return matrix;
};
/**
 * @function parseStringId
 * @description Convert string with 'N-N' format and return number for both coordenates
 * @param {String} id - 'N-N' format
 * @returns {Array} with [ x, y ] format where x & y are numbers
 */
export const parseStringId = (id: string): number[] => {
  if (!/^\d+-\d+$/.test(id)) return [];
  const coordenate = id.split('-');
  const x = Number(coordenate[0]);
  const y = Number(coordenate[1]);
  return [x, y];
};

/**
 * @function getIslands
 * @description - Get matrix groups ('islands')
 * @param {ICell[]} activeCells - Active cells
 * @returns {Array<Array>} - Return an array with arrays for each group ("island") found
 */
export const getIslands = (activeCells: ICell[]): string[][] => {
  const activeIds: string[] = activeCells.map(cell => cell.id);
  let result: string[][] = [];
  activeIds.forEach(p1 => {
    const groups = result.filter(group => group.some(p2 => isFamily(p1, p2)));
    if (groups.length) {
      result = result.filter(r => !groups.some(g => g === r));
      return result.push([...groups.flat(), p1]);
    }
    result.push([p1]);
  });
  return result;
};

export const calculateCellSize = (
  rowCount: number,
  windowWidth: number,
  spacingX: number
): number => {
  const space = windowWidth - spacingX;
  return Number(space / rowCount);
};

/**
 * @function isFamily
 * @param {String} idA - id to compre
 * @param {String} idB - id to compare
 * @returns {Boolean} - boolean value containing if id 'A' is family of id 'B'
 */
export const isFamily = (idA: string, idB: string): boolean => {
  if (idA === idB) return true;
  if (!idA) return false;
  const [x, y] = parseStringId(idA);
  switch (idB) {
    case `${x}-${y - 1}`:
    case `${x}-${y + 1}`:
    case `${x - 1}-${y}`:
    case `${x + 1}-${y}`:
      return true;
    default:
      return false;
  }
};

export const idHasFamilyOnArr = (id: string, arr: string[]): boolean => {
  if (!Array.isArray(arr)) return false;
  for (let idB of arr) {
    if (isFamily(id, idB)) {
      return true;
    }
  }
  return false;
};

export const getActiveCells = (matrix: IMatrix): ICell[] => {
  let active: ICell[] = [];
  for (let row of matrix) {
    for (let cell of row) {
      if (cell.active) {
        active.push(cell);
      }
    }
  }
  return active;
};
/**
 * Returns boolean whether two points are diagonally aligned;
 * @param { String } idA
 * @param { String } idB
 */
export const isDiagonal = (idA: string, idB: string): boolean => {
  const [x, y] = parseStringId(idA);
  return [
    `${x - 1}-${+y - 1}`,
    `${x - 1}-${+y + 1}`,
    `${x + 1}-${+y + 1}`,
    `${x + 1}-${+y - 1}`
  ].some(e => idB === e);
};
/**
 * Returns the edges of a diagonal
 * @param { String } idA
 * @param { String } idB
 */
export const getAristas = (idA: string, idB: string) => {
  const [Ax, Ay] = parseStringId(idA);
  const [Bx, By] = parseStringId(idB);
  const simbolX = Bx - Ax;
  const simbolY = By - Ay;
  return [`${Ax + simbolX}-${Ay}`, `${Ax}-${Ay + simbolY}`];
};
