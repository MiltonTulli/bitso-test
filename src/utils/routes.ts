export default {
  ROOT: '/',
  BOARD: '/board'
};
