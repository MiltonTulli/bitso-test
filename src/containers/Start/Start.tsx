import React from 'react';
import { connect } from 'react-redux';
import { Landing } from '../../components';
import { actions } from '../../modules/world';
import { bindActionCreators } from 'redux';
import { parseStringId } from '../../utils/matrix.utils';

interface StartMapStateToProps {
  saveMatrixDimensions: actions.ActionCreator<{ columns: number; rows: number }>;
}

interface StartProps extends StartMapStateToProps {}

export const Start: React.FunctionComponent<StartProps> = (props: StartProps) => {
  const changeMatrixDimensions = (value: string) => {
    const [columns, rows] = parseStringId(value.replace('x', '-'));
    props.saveMatrixDimensions({ columns, rows });
  };

  return <Landing onDimensionsChange={changeMatrixDimensions} />;
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      saveMatrixDimensions: actions.saveMatrixDimensions
    },
    dispatch
  );

export default connect(null, mapDispatchToProps)(Start);
