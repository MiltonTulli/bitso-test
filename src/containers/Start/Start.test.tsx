import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Start } from './Start';
import { withRouter } from '../../utils/test';

describe('<Start>', () => {
  let props;

  beforeEach(() => {
    props = {
      saveMatrixDimensions: jest.fn()
    };
  });

  it('Should match snapshot', () => {
    const wrapper = renderer.create(withRouter(<Start {...props} />)).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('Should trigger onChange', () => {
    const wrapper = shallow(<Start {...props} />);
    const select = wrapper
      .find('Landing')
      .dive()
      .find('[data-testid="changeMatrixSize"]');
    select.simulate('change', { target: { value: '5x10' } });
    expect(props.saveMatrixDimensions).toHaveBeenCalledWith({ columns: 5, rows: 10 });
  });
});
