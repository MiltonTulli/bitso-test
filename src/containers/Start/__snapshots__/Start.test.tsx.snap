// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Start> Should match snapshot 1`] = `
<div
  className="root-0-2-1"
>
  <section>
    <div
      className="icon-list"
    >
      <i
        className="snes-jp-logo"
      />
    </div>
    <h1
      className="title"
    >
      Bitso js challenge
    </h1>
    <p>
      ~ Milton Tulli ~
       
      <a
        href="https://www.linkedin.com/in/miltontulli/"
      >
        <i
          className="nes-icon linkedin is-small"
        />
      </a>
    </p>
  </section>
  <section>
    <div
      className="nes-container with-title selectContainer-0-2-2"
    >
      <p
        className="title"
      >
        Choose Board Size
      </p>
      <div
        className="nes-select"
      >
        <select
          data-testid="changeMatrixSize"
          defaultValue="-1"
          onChange={[Function]}
        >
          <option
            disabled={true}
            value="-1"
          >
            Select..
          </option>
          <option
            value="5x5"
          >
            5x5
          </option>
          <option
            value="10x5"
          >
            10x5
          </option>
          <option
            value="10x10"
          >
            10x10
          </option>
          <option
            value="15x10"
          >
            15x10
          </option>
          <option
            value="15x15"
          >
            15x15
          </option>
          <option
            value="20x20"
          >
            20x20
          </option>
        </select>
      </div>
    </div>
  </section>
  <section>
    <button
      className="nes-btn is-primary"
      onClick={[Function]}
      type="button"
    >
      Go to App
    </button>
  </section>
</div>
`;
